hudi-timeline-service/src/main/java/org/apache/hudi/timeline/service/handlers/MarkerHandler.java

```diff
@@ -0,0 +1,297 @@
+            .map(fileStatus -> fileStatus.getPath().toString())
+            .collect(Collectors.toList());
+
+        if (markerDirSubPaths.size() > 0) {
+          for (String subPathStr: markerDirSubPaths) {
+            fileSystem.delete(new Path(subPathStr), true);
+          }
+        }
+
+        boolean result = fileSystem.delete(markerDirPath, true);
+        LOG.info("Removing marker directory at " + markerDirPath);
+        return result;
+      }
+    } catch (IOException ioe) {
+      throw new HoodieIOException(ioe.getMessage(), ioe);
+    }
+    return false;
+  }
+
+  private Set<String> getAllMarkersFromFile(String markerDirPath) {
+    LOG.info("Get all markers from " + markerDirPath);
+    Path markersFilePath = new Path(markerDirPath, MARKERS_FILENAME);
+    Set<String> markers = new HashSet<>();
+    try {
+      if (fileSystem.exists(markersFilePath)) {
+        LOG.info("Marker file exists: " + markersFilePath.toString());
+        FSDataInputStream fsDataInputStream = fileSystem.open(markersFilePath);
+        BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(fsDataInputStream, StandardCharsets.UTF_8));
+        markers = bufferedReader.lines().collect(Collectors.toSet());
+        bufferedReader.close();
+        fsDataInputStream.close();
+      } else {
+        LOG.info("Marker file not exist: " + markersFilePath.toString());
+      }
+    } catch (IOException e) {
+      throw new HoodieIOException("Failed to read MARKERS file " + markerDirPath, e);
+    }
+    return markers;
+  }
+
+  private class CreateMarkerCompletableFuture extends CompletableFuture<String> {
+    private Context context;
+    private final String markerDirPath;
+    private final String markerName;
+    private boolean result;
+    private final long startTimeMs;
+
+    public CreateMarkerCompletableFuture(Context context, String markerDirPath, String markerName) {
+      super();
+      this.startTimeMs = System.currentTimeMillis();
+      this.context = context;
+      this.markerDirPath = markerDirPath;
+      this.markerName = markerName;
+      this.result = false;
+    }
+
+    public Context getContext() {
+      return context;
+    }
+
+    public String getMarkerDirPath() {
+      return markerDirPath;
+    }
+
+    public String getMarkerName() {
+      return markerName;
+    }
+
+    public boolean getResult() {
+      return result;
+    }
+
+    public void setResult(boolean result) {
+      LOG.info("*** Request queued for " + (System.currentTimeMillis() - startTimeMs) + " ms");
+      this.result = result;
+    }
+  }
+
+  private class BatchCreateMarkerRunnable implements Runnable {
+
+    @Override
+    public void run() {
+      LOG.info("Start processing create marker requests");
+      long startTimeMs = System.currentTimeMillis();
+      List<CreateMarkerCompletableFuture> futuresToRemove = new ArrayList<>();
+      Set<String> updatedMarkerDirPaths = new HashSet<>();
```

Comment:
for a given commit, can we have multiple entries in this set? in other words, I thought all create requests will have same marker directory path for a given commit.

Reply:
For a single commit, the `updatedMarkerDirPaths` Set should contain only one marker directory since all the creation requests have the same commit instant which is used to construct the marker directory for the commit.

I still use a Set here for multi-commit use case where marker creation requests from different commit are received so the `MarkerHandler` need to differentiate the commits.